File: jksoa-rpc/jksoa-rpc-client/src/test/resources/IPhpGuardService.php
<?php
namespace net\jkcode\jksoa\rpc\example;

class IPhpGuardService {

	// ------------ 方法注解 ------------
	public static $_methodAnnotations = [
		"getUserByNameAsync" => ["net.jkcode.jkguard.annotation.GroupCombine" => ["batchMethod" => "listUsersByNameAsync","reqArgField" => "name","respField" => "","one2one" => "true","flushQuota" => "100","flushTimeoutMillis" => "100"]],
		"listUsersByNameAsync" => ["net.jkcode.jkguard.annotation.RateLimit" => ["permitsPerSecond" => "100.0","stablePeriodSeconds" => "0","warmupPeriodSeconds" => "0"]],
		"getUserWhenRandomException" => ["net.jkcode.jkguard.annotation.CircuitBreak" => ["type" => "CircuitBreakType.EXCEPTION_COUNT","threshold" => "1.0","checkBreakingSeconds" => "5","breakedSeconds" => "5","rateLimit" => "RateLimit(0.0)"],
										"net.jkcode.jkguard.annotation.Metric" => ["bucketCount" => "60","bucketMillis" => "1000","slowRequestMillis" => "10000"]],
		"getUserWhenException" => ["net.jkcode.jkguard.annotation.Degrade" => ["fallbackMethod" => "getUserWhenFallback"]],
		"getUserByIdAsync" => ["net.jkcode.jkguard.annotation.KeyCombine" => []]
	];

	// ------------ php对java调用映射的方法 ------------
	public static function getUserByIdAsync($id){
		return 'java.util.concurrent.CompletableFuture getUserByIdAsync(int)';
	}

	public static function getUserByNameAsync($name){
		return 'java.util.concurrent.CompletableFuture getUserByNameAsync(java.lang.String)';
	}

	public static function listUsersByNameAsync($names){
		return 'java.util.concurrent.CompletableFuture listUsersByNameAsync(java.util.List)';
	}

	public static function getUserWhenException($id){
		return 'java.util.Map getUserWhenException(int)';
	}

	public static function getUserWhenRandomException($id){
		return 'java.util.Map getUserWhenRandomException(int)';
	}

	// ------------ 降级方法 ------------
	public static function getUserWhenFallback($id){
		// TODO: 手动实现降级方法 
	}

}

File: jksoa-rpc/jksoa-rpc-client/src/main/kotlin/net/jkcode/jksoa/rpc/example/IPhpGuardService.kt
package net.jkcode.jksoa.rpc.example

import net.jkcode.jkguard.annotation.*
import net.jkcode.jksoa.common.annotation.RemoteService
import net.jkcode.jkguard.circuit.CircuitBreakType
import java.io.Serializable
import java.util.concurrent.CompletableFuture

/**
 * 给php调用的守护者示例的服务接口
 *
 * @author shijianhang
 * @create 2017-12-15 下午7:37
 **/
@RemoteService(version = 1)
interface IPhpGuardService /*: Remote // rmi协议服务接口 */ {

    // key合并
    @KeyCombine
    fun getUserByIdAsync(id: Int): CompletableFuture<Map<String, Any?>>

    // group合并
    @GroupCombine("listUsersByNameAsync", "name", "", true, 100, 100)
    fun getUserByNameAsync(name: String): CompletableFuture<Map<String, Any?>>

    // group合并后要调用的批量方法
    // 限流
    @RateLimit(100.0)
    fun listUsersByNameAsync(names: List<String>): CompletableFuture<List<Map<String, Any?>>>

    // 降级: 有异常后备方法
    @Degrade(fallbackMethod = "getUserWhenFallback")
    fun getUserWhenException(id: Int): Map<String, Any?>

    // 发送异常时调用的方法, 一般是默认方法
    @JvmDefault
    fun getUserWhenFallback(id: Int): Map<String, Any?> {
        return buildUser(-1, "无名氏")
    }

    // 统计请求数
    @Metric()
    // 断路器
    @CircuitBreak(CircuitBreakType.EXCEPTION_COUNT, 1.0, 5, 5)
    fun getUserWhenRandomException(id: Int): Map<String, Any?>
}

fun buildUser(id: Int, name: String): Map<String, Any> {
    return mapOf("id" to id, "name" to name)
}